pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageRedeliveryTracker.java

```diff
@@ -71,9 +72,13 @@ public void run(Timeout t) throws Exception {
                     }
                     headPartition.clear();
                     redeliveryTimePartitions.addLast(headPartition);
-                    triggerRedelivery(consumerBase);
+                    messageIds = getRedeliveryMessages(consumerBase);
                 } finally {
                     writeLock.unlock();
+                    if (messageIds != null && !messageIds.isEmpty()) {
+                        consumerBase.onAckTimeoutSend(messageIds);
+                        consumerBase.redeliverUnacknowledgedMessages(messageIds);
```

Comment:
Need to unlock first, and then redeliver unacknowledge messages. Otherwise, there will be a deadlock.

Reply:
@shibd Could you please provide more details of the deadlock?

Reply:
Oh, I think you mean you need to change triggerRedelivery -> getRedeliveryMessages because of the deadlock issue right?

Reply:
You can refer to the timing diagram below:

|Time | UnAckedMessageRedeliverTracker timer run | internalPinnedExecutor |
|--| --| --|
|0| [writeLock.lock()](https://github.com/apache/pulsar/blob/a2a73beb52865ddfcef92dab37ad0bcf3bc49fd5/pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageRedeliveryTracker.java#L62) // **first lock**| |
|1| [consumerBase.redeliverUnacknowledgedMessages(messageIds)](https://github.com/apache/pulsar/blob/74b5d505554ea75c81cd004c3e5b61af094848bf/pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageRedeliveryTracker.java#L122)||
|2| ConsumerImpl.redeliverUnacknowledgedMessages()| |
|3| internalRedeliverUnacknowledgedMessages().get(); // block ||
|4|  |internalRedeliverUnacknowledgedMessages()|
|5| | [unAckedMessageTracker.clear();](https://github.com/apache/pulsar/blob/492c7df1579c12701da6a77af1986d3aa7ace840/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1920)|
|6| |writeLock.lock() // **second lock, blocked forever**|

There was no problem before because `redeliverUnacknowledgedMessages` was executed in a `UnAckedMessageRedeliverTracker.timer thread`.

Reply:
Oh, I see.